caffe2/serialize/inline_container.h

```diff
@@ -59,195 +63,271 @@ namespace torch { namespace jit {
 //         been written. We place the variable-length index at the end and do
 //         not put any indicies into the header to fulfill this constraint.
 
+// The serialized model, which contains all the metadata information,
+// should be stored as the last record. Because the size of tensor data is
+// usually stable. As long as the shape and type of the tensor do not change,
+// the size of the data won't change. On the other sied, the size of the
+// serialized model is likely to change, so we store it as the last record, and
+// we don't need to move previous records when updating the model data.
+
 namespace {
-  struct RecordTags {
-    enum {
-      STORAGE = 1,
-      FOOTER = 2,
-    };
-  };
 
-  // Common constants
-  static constexpr uint64_t kFileMagicNumber = 0x314843524f545950L; // PYTORCH1
-  static constexpr uint64_t kFieldAlignment = 64L; // 64 byte alignment supports up to AVX512 for mmap
+enum RecordTags {
+  STORAGE = 1,
+  FOOTER = 2,
+};
+
+// Common constants
+constexpr uint64_t kFileMagicNumber = 0x314843524f545950L; // PYTORCH1
+constexpr uint64_t kFieldAlignment =
+    64L; // 64 byte alignment supports up to AVX512 for mmap
 
-  // Reader-specific constants
-  static constexpr uint64_t kMaxSupportedFileFormatVersion = 0x1L;
+// Reader-specific constants
+constexpr uint64_t kMaxSupportedFileFormatVersion = 0x1L;
 
-  // Writer-specific constants
-  static constexpr uint64_t kFileFormatVersion = 0x1L;
-  static constexpr uint8_t kPadValue = 0xEF;
+// Writer-specific constants
+constexpr uint64_t kFileFormatVersion = 0x1L;
+constexpr char kPadValue = -17; // 0xEF
```

Comment:
why not unsigned char?

Reply:
because ostream write char, and 0xEF is out of range, converting 0xEF to char is implementation based (kind of undefined behavior). When we split .h to .cc, we will solve this problem by using static array.